### PR TITLE
add RSA OAEP SHA2 vectors via custom vector creation

### PR DESCRIPTION
**Pull request by reaperhulk**
This is a work in progress

- [x] Validation
- [x] Proper header on every vector
- [x] Remove the commented out OAEP code (that should be removed in a separate PR that adds tests using these vectors)

refs #1663

### CODE REVIEW

**Comment by chellygel** [comment]
Paul and I spent the evening at PYCON inspecting this code review. He explained it in detail. We were also able to compile and run the verification successfully.

I leave it in the all powerful alex's hands. GLHF

File: docs/development/custom-vectors/rsa-oaep-sha2/Verify_RSA_OAEP_SHA2.java [hunk 1]
```diff
@@ -0,0 +1,413 @@
...
+        m_data = new TestVectorData();
+    }
+
+    public void close() {
+        try {
+            if (m_reader != null) {
+                m_reader.close();
+            }
+            if (m_file_reader != null) {
+                m_file_reader.close();
```
Comment (dreid) [inline docs/development/custom-vectors/rsa-oaep-sha2/Verify_RSA_OAEP_SHA2.java]: You should set the to null after you close them.

File: docs/development/custom-vectors/rsa-oaep-sha2/Verify_RSA_OAEP_SHA2.java [hunk 2]
```diff
@@ -0,0 +1,413 @@
...
+
+        if (line.startsWith(OAEP_EXAMPLE_HEADER)) {
+            // Next example, keep existing keys and load next message
+            loadMessage(m_reader, m_data);
+            return m_data;
+        }
+
+        // otherwise it's a new example
+        if (!line.startsWith(EXAMPLE_HEADER))
+            throw new IOException("Test Header Missing");
```
Comment (dreid) [inline docs/development/custom-vectors/rsa-oaep-sha2/Verify_RSA_OAEP_SHA2.java]: Always use curly braces.

File: docs/development/custom-vectors/rsa-oaep-sha2/Verify_RSA_OAEP_SHA2.java [hunk 3]
```diff
@@ -0,0 +1,413 @@
...
+    private void startNewTest(BufferedReader br) throws IOException {
+        String line = br.readLine();
+        if (!line.startsWith(EXAMPLE))
+            throw new IOException("Example Header Missing");
+    }
+
+    private void loadPublicKey(BufferedReader br, TestVectorData data) throws IOException {
+        String line = br.readLine();
+        if (!line.startsWith(PUB_MODULUS))
+            throw new IOException("Pub Key Modulus Missing");
```
Comment (dreid) [inline docs/development/custom-vectors/rsa-oaep-sha2/Verify_RSA_OAEP_SHA2.java]: Don't abbreviate unecessarily, pub -> public, priv -> private

File: docs/development/custom-vectors/rsa-oaep-sha2/Verify_RSA_OAEP_SHA2.java [hunk 4]
```diff
@@ -0,0 +1,413 @@
...
+        m_cipher = getCipher(m_alg_hash);
+    }
+
+    private Cipher getCipher(SHAHash alg_hash) throws GeneralSecurityException {
+        Cipher cipher = null;
+
+        switch (alg_hash) {
+
+        case SHA1:
+            cipher = Cipher.getInstance("RSA/ECB/OAEPwithSHA1andMGF1Padding", "BC");
```
Comment (dreid) [inline docs/development/custom-vectors/rsa-oaep-sha2/Verify_RSA_OAEP_SHA2.java]: 🙈

File: docs/development/custom-vectors/rsa-oaep-sha2/Verify_RSA_OAEP_SHA2.java [hunk 5]
```diff
@@ -0,0 +1,413 @@
...
+        if (Arrays.equals(java_plaintext, plaintext) == false) {
+            throw new Exception("Verification failure - plaintext does not match after decryption.");
+        }
+    }
+
+    public static void main(String[] args) {
+        Security.addProvider(new BouncyCastleProvider());
+
+        // assume current directory if no path given on command line
+        String vector_path = "./";
```
Comment (dreid) [inline docs/development/custom-vectors/rsa-oaep-sha2/Verify_RSA_OAEP_SHA2.java]: Instructions should tell you how to compile and run this from the root, vectors are in a known location from the root, that should be the default. Once that is the default I don't think there is a good reason to bother with the command line argument?

File: docs/development/custom-vectors/rsa-oaep-sha2/Verify_RSA_OAEP_SHA2.java [hunk 6]
```diff
@@ -0,0 +1,413 @@
...
+
+        line = br.readLine();
+        if (!line.startsWith(ENCRYPTION))
+            throw new IOException("Ciphertext Missing");
+        data.ciphertext = unhexlify(br.readLine());
+    }
+
+}
+
+public class Verify_RSA_OAEP_SHA2 {
```
Comment (dreid) [inline docs/development/custom-vectors/rsa-oaep-sha2/Verify_RSA_OAEP_SHA2.java]: Java Class names do not typically include underscores.

File: docs/development/custom-vectors/rsa-oaep-sha2/generate_rsa_oaep_sha2.py [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+        output.append(b"# Public key")
+        output.append(b"# Modulus:")
+        output.append(format(pn.public_numbers.n, "x"))
+        output.append(b"# Exponent:")
+        output.append(format(pn.public_numbers.e, "x"))
+        output.append(b"# Private key")
+        output.append(b"# Modulus:")
+        output.append(format(pn.public_numbers.n, "x"))
+        output.append(b"# Public exponent:")
+        output.append(format(pn.public_numbers.e, "x"))
```
Comment (dreid) [inline docs/development/custom-vectors/rsa-oaep-sha2/generate_rsa_oaep_sha2.py]: These are already in the file? Can… we just get them from the public section? or omit the public section?

File: docs/development/custom-vectors/rsa-oaep-sha2.rst [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+    $ export CLASSPATH=~/Downloads/bcprov-jdk15on-154.jar:./
+    $ javac Verify_RSA_OAEP_SHA2.java
+
+Finally, run the program with the path to the SHA-2 vectors:
+
+.. code-block:: console
+
+    $ java Verify_RSA_OAEP_SHA2 <path to test vector files>
+
+.. _`Bouncy Castle`: https://www.bouncycastle.org/
```
Comment (dreid) [inline docs/development/custom-vectors/rsa-oaep-sha2.rst]: Link to the release page: https://www.bouncycastle.org/latest_releases.html

File: docs/development/custom-vectors/rsa-oaep-sha2.rst [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+
+Download the latest `Bouncy Castle`_ JAR.  Initial verification was performed
+using ``bcprov-jdk15on-154.jar``.
+
+Set the CLASSPATH to include the Bouncy Castle jar and the path to
+Verify_RSA_OAEP_SHA2.java and compile the program.
+
+.. code-block:: console
+
+    $ export CLASSPATH=~/Downloads/bcprov-jdk15on-154.jar:./
```
Comment (dreid) [inline docs/development/custom-vectors/rsa-oaep-sha2.rst]: Don't mess with the user's environment, pass -classpath to stuff.

File: docs/development/custom-vectors/rsa-oaep-sha2.rst [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+    $ javac Verify_RSA_OAEP_SHA2.java
```
Comment (dreid) [inline docs/development/custom-vectors/rsa-oaep-sha2.rst]: pass `-d .` so the class file ends up in `.`

**Comment by dreid** [comment]
Add *.class to .gitignore probably.

File: docs/development/custom-vectors/rsa-oaep-sha2.rst (hunk 10 above)
Comment (dreid) [inline docs/development/custom-vectors/rsa-oaep-sha2.rst]: Or maybe put them in some predictable ignored directory.

**Pull request merged**